### ISSUE TITLE
[Bug]: First invocation of plt.plot() resets mpl.rcParams['figure.dpi']

### ISSUE THREAD

**Issue by emazep** [matplotlib/matplotlib#23007]
### Bug summary

First invocation of `plt.plot()` resets `mpl.rcParams['figure.dpi']`.

### Code for reproduction

```python
import matplotlib as mpl
import matplotlib.pyplot as plt

mpl.rcParams['figure.dpi'] = 300

print(mpl.rcParams['figure.dpi'])
plt.plot()
print(mpl.rcParams['figure.dpi'])
```

### Actual outcome

300.0
72.0

### Expected outcome

300.0
300.0

### Additional information

This happens only at the very first invocation of `plt.plot()`, while on subsequent `plt.plot()` calls (tested only in Jupyter really), `mpl.rcParams['figure.dpi']` will no longer be overwritten.

Other `rcParams` values seem not to be affected (but I've not tested them all).

I have never observed this behavior in any of the previous of matplotlib versions, so the bug must have been introduced in the latest version (3.5.2).

### Operating system

Windows 10

### Matplotlib Version

3.5.2

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

Python 3.9.12

### Jupyter version

3.4.0

### Installation

conda

**Comment by jklymak** [comment]
We really need to solidify if rcParams track current state or not.  I pretty strongly feel they should not track current state - the figure or its canvas tracks that.  I think rcParams should be read-only from the library's point of view.

**Comment by anntzer** [comment]
Agreed, but AFAICT this does not come from us, but from IPython's integration at https://github.com/ipython/matplotlib-inline/blob/10234a68898f353add344d1c73122df53839fe1b/matplotlib_inline/config.py#L38-L54 and https://github.com/ipython/matplotlib-inline/blob/10234a68898f353add344d1c73122df53839fe1b/matplotlib_inline/backend_inline.py#L182

**Comment by emazep** [comment]
I can confirm that it does not happen in a stand-alone script (and that it has only been tested in Jupyter lab 3.4.0).

**Comment by yut23** [comment]
This has newly appeared in 3.5.2 due to #22005 making IPython's integration code run at the first use instead of at import. Before, it would just override the relevant values set in the matplotlibrc file and not interfere with any runtime changes to rcParams.

**Comment by tacaswell** [comment]
> This has newly appeared in 3.5.2 due to #22005 making IPython's integration code run at the first use instead of at import

Sigh, making this lazier always produces interesting unintended consequences.

> We really need to solidify if rcParams track current state or not. I pretty strongly feel they should not track current state ...

I think the only rcParam that we mutate for the user is the save-folder location (which the user has to opt-into the behavior).  We had a long discussion about that and concluded it was better to just leave that as it is rather than go through the effort of deprecating that behavior and moving the global state someplace else for the sake of aesthetics.

**Comment by neon-ninja** [comment]
I'm having this problem too, with `plt.rcParams['figure.figsize']`. Rolling back to matplotlib==3.5.1 with `pip install matplotlib==3.5.1` fixes it

**Comment by tacaswell** [comment]
Has anyone reported this to matplotlib-inline?

**Comment by neon-ninja** [comment]
This PR in matplotlib-inline would fix the issue - https://github.com/ipython/matplotlib-inline/pull/14/files

**Comment by mikofski** [comment]
I was having a similar issue with `rcParams['figure.figsize']`, see https://discourse.matplotlib.org/t/bug-rcparams-doesnt-work-first-time-v3-5-2/22853, downgrading to mpl-3.5.1 solved the issue. IDK if it's the same?

**Comment by timhoffm** [comment]
> This PR in matplotlib-inline would fix the issue - [ipython/matplotlib-inline#14](https://github.com/ipython/matplotlib-inline/pull/14)

was released in matplotlib-inline 0.1.4 (Aug. 14, 2022). So I think we can close this.
